src/backtracking/rat_in_maze.rs

```diff
@@ -0,0 +1,250 @@
+                    }
+                }
+            )*
+        }
+    }
+
+    test_find_path_in_maze! {
+        maze_with_solution_5x5: 0, 0, &[
+            vec![1, 0, 1, 0, 0],
+            vec![1, 1, 0, 1, 0],
+            vec![0, 1, 1, 1, 0],
+            vec![0, 0, 0, 1, 1],
+            vec![0, 1, 0, 0, 1],
+        ], Some(vec![
+            vec![1, 0, 0, 0, 0],
+            vec![1, 1, 0, 0, 0],
+            vec![0, 1, 1, 1, 0],
+            vec![0, 0, 0, 1, 1],
+            vec![0, 0, 0, 0, 1],
+        ]),
+        maze_with_solution_6x6: 0, 0, &[
+            vec![1, 0, 1, 0, 1, 0],
+            vec![1, 1, 0, 1, 0, 1],
+            vec![0, 1, 1, 1, 1, 0],
+            vec![0, 0, 0, 1, 1, 1],
+            vec![0, 1, 0, 0, 1, 0],
+            vec![1, 1, 1, 1, 1, 1],
+        ], Some(vec![
+            vec![1, 0, 0, 0, 0, 0],
+            vec![1, 1, 0, 0, 0, 0],
+            vec![0, 1, 1, 1, 1, 0],
+            vec![0, 0, 0, 0, 1, 0],
+            vec![0, 0, 0, 0, 1, 0],
+            vec![0, 0, 0, 0, 1, 1],
+        ]),
+        maze_with_solution_8x8: 0, 0, &[
+            vec![1, 0, 0, 0, 0, 0, 0, 1],
+            vec![1, 1, 0, 1, 1, 1, 0, 0],
+            vec![0, 1, 1, 1, 0, 0, 0, 0],
+            vec![0, 0, 0, 1, 0, 1, 1, 0],
+            vec![0, 1, 0, 1, 1, 1, 0, 1],
+            vec![1, 0, 1, 0, 0, 1, 1, 1],
+            vec![0, 0, 1, 1, 1, 0, 1, 1],
+            vec![1, 1, 1, 0, 1, 1, 1, 1],
+        ], Some(vec![
+            vec![1, 0, 0, 0, 0, 0, 0, 0],
+            vec![1, 1, 0, 0, 0, 0, 0, 0],
+            vec![0, 1, 1, 1, 0, 0, 0, 0],
+            vec![0, 0, 0, 1, 0, 0, 0, 0],
+            vec![0, 0, 0, 1, 1, 1, 0, 0],
+            vec![0, 0, 0, 0, 0, 1, 1, 1],
+            vec![0, 0, 0, 0, 0, 0, 0, 1],
+            vec![0, 0, 0, 0, 0, 0, 0, 1],
+        ]),
+        maze_without_solution_4x4: 0, 0, &[
+            vec![1, 0, 0, 0],
+            vec![1, 1, 0, 0],
+            vec![0, 1, 1, 0],
+            vec![0, 0, 0, 1],
+        ], None::<Vec<Vec<usize>>>,
+        maze_no_solution_possible_loops: 0, 0, &[
+            vec![1, 1, 1, 1],
+            vec![1, 1, 1, 1],
+            vec![1, 1, 1, 0],
+            vec![1, 1, 0, 1],
+        ], None::<Vec<Vec<usize>>>,
+        maze_with_start_outside: 5, 0, &[
+            vec![1, 0, 0, 0],
+            vec![1, 1, 0, 0],
+            vec![0, 1, 1, 0],
+            vec![0, 0, 1, 1],
+        ], None::<Vec<Vec<usize>>>,
+        unproper_maze_with_solution: 0, 0, &[
+            vec![1],
+            vec![1, 1],
+            vec![1, 1, 1],
+            vec![1, 1, 1, 1]
+        ], Some(vec![
+            vec![1, 0, 0, 0],
+            vec![1, 1, 0, 0],
+            vec![0, 1, 1, 0],
+            vec![0, 0, 1, 1]
+        ]),
+        unproper_maze_no_solution_possible_loops: 0, 0, &[
+            vec![1, 1, 1, 1],
+            vec![1, 1],
+            vec![1, 1, 0],
+            vec![0, 0, 1],
+        ], None::<Vec<Vec<usize>>>,
```

Comment:
One more thing: could you please add a test case with input not being a square, for example `2x3` like input?

Reply:
I will try